sgx/common/src/retrieve_key_protocol.c

```diff
@@ -0,0 +1,1164 @@
+  uint16_t temp_val = 0;
+  unsigned char *buf_ptr = pt_msg.body;
+
+  // insert KMIP 'get key' response size
+  temp_val = htobe16((uint16_t) kmip_response->size);
+  memcpy(buf_ptr, &temp_val, 2);
+  buf_ptr += 2;
+
+  // insert KMIP 'get key' response bytes
+  memcpy(buf_ptr, kmip_response->buffer, kmip_response->size);
+
+  // append signature to unencrypted 'Key Response' message
+  if (EXIT_SUCCESS != append_msg_signature(server_sign_key, &pt_msg))
+  {
+    kmyth_sgx_log(LOG_ERR, "error appending message signature");
+    kmyth_clear_and_free(pt_msg.body, pt_msg.hdr.msg_size);
+    return EXIT_FAILURE;
+  }
+
+  // if output message buffer allocated, free and set to NULL
+  if (msg_out->body != NULL)
+  {
+    free(msg_out->body);
+    msg_out->hdr.msg_size = 0;
+    msg_out->body = NULL;
+  }
+
+  // encrypt signed 'Key Response' message using the specified key
+  if (EXIT_SUCCESS != aes_gcm_encrypt(msg_enc_key->buffer,
+                                      msg_enc_key->size,
+                                      pt_msg.body,
+                                      pt_msg.hdr.msg_size,
+                                      &(msg_out->body),
+                                      (size_t *) &(msg_out->hdr.msg_size)))
+  {
+    kmyth_sgx_log(LOG_ERR, "failed to encrypt the 'Key Response' message");
+    kmyth_clear_and_free(pt_msg.body, pt_msg.hdr.msg_size);
+    return EXIT_FAILURE;
+  }
+  kmyth_clear_and_free(pt_msg.body, pt_msg.hdr.msg_size);
+
+  return EXIT_SUCCESS;
+}
+
+/*****************************************************************************
+ * parse_key_response_msg()
+ ****************************************************************************/
+int parse_key_response_msg(X509 * server_sign_cert,
+                           ByteBuffer * msg_dec_key,
+                           ECDHMessage * msg_in,
+                           ByteBuffer * kmip_response)
+{
+  // decrypt message using input message decryption key
+  ECDHMessage pt_msg = { { 0 }, NULL };
+
+  if (EXIT_SUCCESS != aes_gcm_decrypt(msg_dec_key->buffer,
+                                      msg_dec_key->size,
+                                      msg_in->body,
+                                      msg_in->hdr.msg_size,
+                                      &(pt_msg.body),
+                                      (size_t *) &(pt_msg.hdr.msg_size)))
+  {
+    kmyth_sgx_log(LOG_ERR, "failed to decrypt the 'Key Response' message");
+    if (pt_msg.body != NULL)
+    {
+      kmyth_clear_and_free(pt_msg.body, pt_msg.hdr.msg_size);
+    }
+    return EXIT_FAILURE;
+  }
+
+  // parse message body fields into variables
+  int buf_index = 0;
+
+  // get size (in bytes) of KMIP 'get key' request field
+  kmip_response->size = (pt_msg.body)[buf_index] << 8;
+  kmip_response->size += (pt_msg.body)[buf_index+1];
+  buf_index += 2;
+
+  // get KMIP 'get key' response bytes
+  kmip_response->buffer = malloc(kmip_response->size);
```

Comment:
`kmip_response->buffer` should be freed if an error occurs, and `kmip_response` should be cleared.

Reply:
Attempted to fix this in commit [bfa5d11](https://github.com/NationalSecurityAgency/kmyth/pull/168/commits/bfa5d110d0039e66d371b9f74499e795b1f8c7f1).